pkg/operator/awsTagController.go

```diff
@@ -0,0 +1,324 @@
+	infraTagSet := make(map[string]string)
+	mergePlatformStatusTags(infra, infraTagSet)
+	klog.V(5).Infof("tags read from Infrastructure resource: %v", infraTagSet)
+
+	// Create a driver with the current configuration
+	ctx := context.Background()
+	driver := s3.NewDriver(ctx, cr.Spec.Storage.S3, &c.listers.StorageListers, c.featureGateAccessor)
+
+	s3TagSet, err := driver.GetStorageTags()
+	if err != nil {
+		klog.Errorf("failed to fetch storage tags: %v", err)
+		return err
+	}
+	klog.V(5).Infof("tags read from storage resource: %v", s3TagSet)
+
+	tagUpdatedCount := compareS3InfraTagSet(s3TagSet, infraTagSet)
+	if tagUpdatedCount > 0 {
+		if err := driver.PutStorageTags(s3TagSet); err != nil {
+			klog.Errorf("failed to update/delete tagset of %s s3 bucket: %v", driver.ID(), err)
+			c.event.Warningf("UpdateAWSTags",
+				"Failed to update/delete tagset of %s s3 bucket", driver.ID())
+		}
+		klog.Infof("successfully updated/deleted %d tags, tagset: %+v", tagUpdatedCount, s3TagSet)
+		c.event.Eventf("UpdateAWSTags",
+			"Successfully updated/deleted tagset of %s s3 bucket", driver.ID())
+	}
+
+	return nil
+}
+
+// mergePlatformStatusTags is for reading and merging user tags present in both
+// Platform Status of Infrastructure config.
+// There could be scenarios(upgrade, user deletes) where user tags could be missing
+// from the Platform Status, hence using Status too to avoid said scenarios.
+// If a tag exists in Status.
+func mergePlatformStatusTags(infra *configv1.Infrastructure, infraTagSet map[string]string) {
+	for _, statusTags := range infra.Status.PlatformStatus.AWS.ResourceTags {
+		if err := validateUserTag(statusTags.Key, statusTags.Value); err != nil {
+			klog.Warningf("validation failed for tag(%s:%s): %v", statusTags.Key, statusTags.Value, err)
+			continue
+		}
+		infraTagSet[statusTags.Key] = statusTags.Value
+	}
+}
+
+// compareS3InfraTagSet is for comparing the tags obtained from S3 bucket and Infrastructure CR
+// to find if any new tags have been deleted, added or existing tags modified.
+func compareS3InfraTagSet(s3TagSet map[string]string, infraTagSet map[string]string) (tagUpdatedCount int) {
+	for key, value := range infraTagSet {
+		// If a tag is value is empty, it's marked for deletion
+		// and is deleted from the list obtained from S3 bucket
+		if value == "" {
+			klog.V(5).Infof("%s tag will be deleted", key)
+			delete(s3TagSet, key)
+			tagUpdatedCount++
+			continue
+		}
+		val, ok := s3TagSet[key]
+		if !ok || val != value {
+			klog.V(5).Infof("%s tag will be added/updated with value %s", key, value)
+			s3TagSet[key] = value
+			tagUpdatedCount++
+		}
+	}
+	return
+}
+
+// validateUserTag is for validating the user defined tags in Infrastructure CR
+func validateUserTag(key, value string) error {
+	if !tagKeyRegex.MatchString(key) {
+		return fmt.Errorf("key has invalid characters or length")
+	}
+	if strings.EqualFold(key, "Name") {
+		return fmt.Errorf("key cannot be customized by user")
```

Comment:
So `key` cannot be `"Name"`? Why can't the user customize it? I thought that was the point of this feature 🤔

Reply:
I think this is required I referred this from the installer code https://github.com/openshift/installer/blob/master/pkg/types/aws/validation/platform.go#L64-L91 

CC: @TrilokGeer